File: sprint1/Java/lesson7code6.java
public static Pair<List<Integer>, int[]> getLeastPrimesLinear(int n) {
    int[] lp = new int[n + 1];
    List<Integer> primes = new ArrayList<>();
    for (int i = 2; i <= n; i++) {
        if (lp[i] == 0) {
            lp[i] = i;
            primes.add(i);
        }
        for (int j = 0; j < primes.size(); j++) {
            int p = primes.get(j);
            int x = p * i;
            if (p > lp[i] || x > n) {
                break;
            }
            lp[x] = p;
        }
    }
    return new Pair<>(primes, lp);
}

File: sprint1/kotlin/lesson7code6.kt
fun getLeastPrimesLinear(n: Int): Pair<IntArray, IntArray> {
    val lp = IntArray(n + 1) { 0 }
    val primes = mutableListOf<Int>()
    for (i in 2..n) {
        if (lp[i] == 0) {
            lp[i] = i
            primes.add(i)
        }
        for (p in primes) {
            val x = p * i
            if (p > lp[i] || x > n) {
                break
            }
            lp[x] = p
        }
    }
    return primes.toIntArray() to lp
}
